pkg/raw/raw_exec.go

```diff
@@ -5,44 +5,98 @@ package raw
 
 import (
 	"fmt"
-	"os/exec"
-	"strconv"
+	"io"
+	"os"
 
+	"github.com/go-logr/logr"
 	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
 )
 
-type Exec struct {
-}
+type Exec struct{}
+
+const filePerm = 0660
 
 func (Exec) Create(filename string, opts ...CreateOption) error {
 	o := &CreateOptions{}
 	o.ApplyOptions(opts)
-	var seek string
-
-	if o.SourceFile == "" && o.Size == nil {
-		return fmt.Errorf("must specify Size when creating without source file")
-	}
+	log := ctrl.Log.WithName("raw-disk")
 
-	ofilename := "of=" + filename
-	if o.Size != nil {
-		seek = "seek=" + strconv.FormatInt(*o.Size, 10) // TODO: verify if the disk size is proper, else G as a suffix
-	}
 	if o.SourceFile == "" {
-		outMsgDD, err := exec.Command("dd", "if=/dev/zero", ofilename, "bs=1", "count=0", seek).Output()
+		if o.Size == nil {
+			return fmt.Errorf("must specify Size when creating without source file")
+		}
+		seek := *o.Size
+		// Position the file cursor one byte before the desired seek position to write a single byte,
+		// to ensure that data is written at the exact byte position specified by seek.
+		err := createEmptyFileWithSeek(log, filename, seek-1)
 		if err != nil {
-			return fmt.Errorf("failed creating the ephemeral disk: %s, exit error %w", string(outMsgDD), err)
+			return fmt.Errorf("failed creating the empty ephemeral disk at %s: %w", filename, err)
 		}
 	} else {
-		ifilename := "if=" + o.SourceFile
-		outMsgDD, err := exec.Command("dd", ifilename, ofilename).Output()
+		err := copyFile(log, o.SourceFile, filename)
 		if err != nil {
-			return fmt.Errorf("failed creating the image for virtual disk: %s, exit error %w", string(outMsgDD), err)
+			return fmt.Errorf("failed creating virtual disk image, source: %s, destination: %s: %w", o.SourceFile, filename, err)
 		}
 	}
 
 	return nil
 }
 
+func createEmptyFileWithSeek(log logr.Logger, filename string, seek int64) error {
+	log = log.WithName("createEmptyFileWithSeek").WithValues("filename", filename)
```

Comment:
Why do we need here a new logger if no logs except one error is being logged? Maybe just attach the filename to the logger while creating it in the `Create(filename string, opts ...CreateOption) error {` function.

Reply:
Added the filename in the log itself.

As we are dealing with two files in the overall program i.e source file and destination file, attaching the filename to the logger while creating it in the `Create(filename string, opts ...CreateOption) error {` function, could lead to confusion in debugging and maintenance.

Keeping the WithName method, as having it for each component will help in providing clear and identifiable log messages, which is generally preferred for better code readability and maintainability.

Reply:
logger is primary for additional logs in Close functions. These errors aren't fatal but they can be problematic for long run apps. So we log them minimally.

Reply:
logging them is totally fine. but it dosen't really make sense to create a new logger for every function. This code belongs to the `Create` function and with a meaningful error the logger passed to `createEmptyFileWithSeek(log logr.Logger` can easily used.

Reply:
sorry i didn't understand your post properly. Filename can be added in Create function. @so-sahu we don't need add target filename because logic is hardcoded here, so i don't see big benefits with logging of destination file. But i recommend keep name of function somewhere in logs for better debugging.

Reply:
Implemented the change, as discussed offline.